core/base/v7/inc/ROOT/RError.hxx

```diff
@@ -0,0 +1,202 @@
+class RStatusTypeBool : public RStatusType<bool> {
+public:
+   RStatusTypeBool() = default;
+   explicit RStatusTypeBool(bool value) : RStatusType<bool>(value) {}
+   bool IsError() const { return fValue == false; }
+   void SetError() { fValue = false; }
+};
+
+// clang-format off
+/**
+\class ROOT::Experimental::Detail::RStatusTypeSyscall
+\ingroup Base
+\brief For system calls that return 0 (or a meaningful integer) on success
+*/
+// clang-format on
+class RStatusTypeSyscall : public RStatusType<int> {
+public:
+   RStatusTypeSyscall() = default;
+   explicit RStatusTypeSyscall(int value) : RStatusType<int>(value) {}
+   bool IsError() const { return fValue < 0; }
+   void SetError() { fValue = -1; }
+};
+
+// clang-format off
+/**
+\class ROOT::Experimental::Detail::RStatusBase
+\ingroup Base
+\brief Type-independent logic of the RStatus wrapper
+*/
+// clang-format on
+class RStatusBase {
+protected:
+   static thread_local bool fgThrowInstantExceptions; /* true by default */
+public:
+   static void SetThrowInstantExceptions(bool value) { fgThrowInstantExceptions = value; }
+};
+
+// clang-format off
+/**
+\class ROOT::Experimental::Detail::RStatus
+\ingroup Base
+\brief Wraps a return value such that unchecked error states trigger an exception
+
+The Status wrapper itself is movable but not copyable to prevent multiple exceptions from being thrown.
+The wrapper is also only stack allocatable because throwing an exception depends on the object going out of scope.
+
+If the class should be used like a C return type, it has to be cleared by a call to IsError() or IsValid() or
+ClearError().  Otherwise, an error state will throw an exception nevertheless when the object goes out of scope.
+*/
+// clang-format on
+template <typename T> // Has to be RStatusType
+class RStatus : public RStatusBase {
+   T fStatus;
+   bool fIsChecked = false;
+
+public:
+   // Named constructor for error cases
+   static RStatus Fail(const std::string &why)
+   {
+      RStatus status;
+      status.MayThrow(why);
+      return status;
+   }
+
+   RStatus() = default;
+   RStatus(const RStatus &other) = delete;
+   RStatus(RStatus &&other) = default;
+   RStatus &operator =(const RStatus &other) = delete;
+   RStatus &operator =(RStatus &&other) = default;
+
+   // Construct from return value
+   explicit RStatus(const typename T::ValueType_t &value) : fStatus(value) {}
+   // Assign a return value
+   RStatus &operator =(const typename T::ValueType_t &value) { this->fStatus = value; }
+
+   ~RStatus() noexcept(false)
+   {
+      if (!fIsChecked && fStatus.IsError()) {
+         // Prevent from throwing if the object is deconstructed in the course of stack unwinding for another exception
+#if __cplusplus >= 201703L
+         if (std::uncaught_exceptions() == 0)
+#else
+         if (!std::uncaught_exception())
+#endif
+         {
+            throw RException("unchecked error");
+         }
+      }
+   }
+
+   void MayThrow(const std::string &why)
```

Comment:
The name lead to think that this routine would most often throw (that is contradicted by the R__unlikely :) ).
What is the semantic/use-case of the function?

Reply:
`MayThrow` is supposed to transition into the error state. If instant exceptions are turned on, an exception will be thrown right away.  Otherwise, the RStatus object expects to be checked and failure to do so will throw an exception on destruction.